From 50d0fc5a265b63fe14eac8e82560012ad112e4b7 Mon Sep 17 00:00:00 2001
From: Gris Ge <fge@redhat.com>
Date: Wed, 15 Oct 2025 16:06:19 +0800
Subject: [PATCH 1/2] Fix PSK authentication when leftid starts with `@`

When `leftid` starts with `@`, the
`/etc/ipsec.d/ipsec-<$connection_uuid>.secrets` file created will
contains content like:

```
@@cli-a.example.org: PSK "LONG_PSK_STRING"
```

To fix that issue, change both `nm_libreswan_get_ipsec_conf()` and
`nm_libreswan_config_psk_write()` to accept sanitized NmSettingVPN only.

The `nm_libreswan_config_psk_write()` will not adding any leading `@`
anymore since the leftid is already sanitized.

Signed-off-by: Gris Ge <fge@redhat.com>
---
 shared/test-utils.c        | 65 ++++++++++++++++++++++++++------------
 shared/utils.c             | 14 +++-----
 shared/utils.h             |  3 ++
 src/nm-libreswan-service.c |  8 ++---
 4 files changed, 57 insertions(+), 33 deletions(-)

diff --git a/shared/test-utils.c b/shared/test-utils.c
index 16a96d1..8bf5888 100644
--- a/shared/test-utils.c
+++ b/shared/test-utils.c
@@ -30,11 +30,14 @@ test_config_write (void)
 {
 	GError *error = NULL;
 	NMSettingVpn *s_vpn;
+	NMSettingVpn *s_vpn_sanitized;
 	char *str;
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "con_name", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "con_name", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
 	                 "conn con_name\n"
@@ -55,11 +58,14 @@ test_config_write (void)
 	                 " modecfgpull=yes\n");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
 	nm_setting_vpn_add_data_item (s_vpn, "dhgroup", "ignored");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "con_name", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "con_name", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
 	                 "conn con_name\n"
@@ -81,13 +87,16 @@ test_config_write (void)
 
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "ikev2", "insist");
 	nm_setting_vpn_add_data_item (s_vpn, "leftcert", "LibreswanClient");
 	nm_setting_vpn_add_data_item (s_vpn, "leftid", "%fromcert");
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn,
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized,
 	                                   "f0008435-07af-4836-a53d-b43e8730e68f",
 			                   NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
@@ -108,13 +117,16 @@ test_config_write (void)
 	                 " modecfgpull=yes\n");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "ikev2", "insist");
 	nm_setting_vpn_add_data_item (s_vpn, "leftrsasigkey", "hello");
 	nm_setting_vpn_add_data_item (s_vpn, "rightrsasigkey", "world");
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "conn", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
 	                 "conn conn\n"
@@ -131,10 +143,13 @@ test_config_write (void)
 	                 " modecfgpull=yes\n");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
-	str = nm_libreswan_get_ipsec_conf (3, s_vpn,
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (3, s_vpn_sanitized,
 	                                   "my_con",
 			                   "/foo/bar/ifupdown hello 123 456",
 	                                   TRUE, FALSE, &error);
@@ -161,6 +176,7 @@ test_config_write (void)
 	                 " nm-configured=yes");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "ikev2", "insist");
@@ -170,7 +186,9 @@ test_config_write (void)
 	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "false");
 	nm_setting_vpn_add_data_item (s_vpn, "leftsendcert", "always");
 	nm_setting_vpn_add_data_item (s_vpn, "rightca", "%same");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "conn", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
                          "conn conn\n"
@@ -182,11 +200,14 @@ test_config_write (void)
                          " rightca=\"%same\"\n");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
 	nm_setting_vpn_add_data_item (s_vpn, "esp", "aes_gcm256");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "con_name", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "con_name", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
 	                 "conn con_name\n"
@@ -209,11 +230,14 @@ test_config_write (void)
 	                 " modecfgpull=yes\n");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
 	nm_setting_vpn_add_data_item (s_vpn, "vendor", "Cisco");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "con_name", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
+	g_assert_no_error (error);
+	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "con_name", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
 	                 "conn con_name\n"
@@ -235,53 +259,54 @@ test_config_write (void)
 	                 " modecfgpull=yes\n");
 	g_free (str);
 	g_object_unref (s_vpn);
+	g_object_unref (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
-	g_assert_null (str);
 	g_clear_error (&error);
 	g_object_unref (s_vpn);
+	g_assert_null (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
 	nm_setting_vpn_add_data_item (s_vpn, "ikev2", "hello world");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
-	g_assert_null (str);
 	g_clear_error (&error);
 	g_object_unref (s_vpn);
+	g_assert_null (s_vpn_sanitized);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12\n13.14");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
-	g_assert_null (str);
+	g_assert_null (s_vpn_sanitized);
 	g_clear_error (&error);
 	g_object_unref (s_vpn);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "rightcert", "\"cert\"");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
-	g_assert_null (str);
+	g_assert_null (s_vpn_sanitized);
 	g_clear_error (&error);
 	g_object_unref (s_vpn);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "false");
 	nm_setting_vpn_add_data_item (s_vpn, "rightcert", "\"cert\"");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
-	g_assert_null (str);
+	g_assert_null (s_vpn_sanitized);
 	g_clear_error (&error);
 	g_object_unref (s_vpn);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "false");
-	str = nm_libreswan_get_ipsec_conf (4, s_vpn, "conn", NULL, FALSE, TRUE, &error);
+	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
-	g_assert_null (str);
+	g_assert_null (s_vpn_sanitized);
 	g_clear_error (&error);
 	g_object_unref (s_vpn);
 }
diff --git a/shared/utils.c b/shared/utils.c
index ac735a5..8dab313 100644
--- a/shared/utils.c
+++ b/shared/utils.c
@@ -384,7 +384,7 @@ check_val (const char *val, gboolean allow_spaces, GError **error)
 	return TRUE;
 }
 
-static NMSettingVpn *
+NMSettingVpn *
 sanitize_setting_vpn (NMSettingVpn *s_vpn,
                       GError **error)
 {
@@ -458,34 +458,30 @@ sanitize_setting_vpn (NMSettingVpn *s_vpn,
 
 char *
 nm_libreswan_get_ipsec_conf (int ipsec_version,
-                             NMSettingVpn *s_vpn,
+                             NMSettingVpn *s_vpn_sanitized,
                              const char *con_name,
                              const char *leftupdown_script,
                              gboolean openswan,
                              gboolean trailing_newline,
                              GError **error)
 {
-	gs_unref_object NMSettingVpn *sanitized = NULL;
 	nm_auto_free_gstring GString *ipsec_conf = NULL;
 	const char *val;
 	int i;
 
-	g_return_val_if_fail (NM_IS_SETTING_VPN (s_vpn), NULL);
+	g_return_val_if_fail (NM_IS_SETTING_VPN (s_vpn_sanitized), NULL);
 	g_return_val_if_fail (!error || !*error, NULL);
 	g_return_val_if_fail (con_name && *con_name, NULL);
 
 	if (!check_val (con_name, FALSE, error))
 		return NULL;
 
-	sanitized = sanitize_setting_vpn (s_vpn, error);
-	if (!sanitized)
-		return NULL;
-
 	ipsec_conf = g_string_sized_new (1024);
 	g_string_append_printf (ipsec_conf, "conn %s\n", con_name);
 
 	for (i = 0; params[i].name != NULL; i++) {
-		val = nm_setting_vpn_get_data_item (sanitized, params[i].name);
+		val = nm_setting_vpn_get_data_item (s_vpn_sanitized,
+						    params[i].name);
 		if (val == NULL)
 			continue;
 
diff --git a/shared/utils.h b/shared/utils.h
index 67718e9..7832966 100644
--- a/shared/utils.h
+++ b/shared/utils.h
@@ -64,4 +64,7 @@ const char *nm_libreswan_find_helper_libexec (const char *progname, GError **err
 gboolean nm_libreswan_parse_subnets (const char *str, GPtrArray *arr, GError **error);
 char *nm_libreswan_normalize_subnets (const char *str, GError **error);
 
+NMSettingVpn *sanitize_setting_vpn (NMSettingVpn *s_vpn, GError **error);
+
+
 #endif /* __UTILS_H__ */
diff --git a/src/nm-libreswan-service.c b/src/nm-libreswan-service.c
index 7987ada..58ada03 100644
--- a/src/nm-libreswan-service.c
+++ b/src/nm-libreswan-service.c
@@ -522,8 +522,7 @@ nm_libreswan_config_psk_write (NMSettingVpn *s_vpn,
 		/* nm_libreswan_get_ipsec_conf() in _connect_common should've checked these. */
 		g_return_val_if_fail (strchr (leftid, '"') == NULL, FALSE);
 		g_return_val_if_fail (strchr (leftid, '\n') == NULL, FALSE);
-
-		secrets = g_strdup_printf ("@%s: PSK \"%s\"", leftid, psk);
+		secrets = g_strdup_printf ("%s: PSK \"%s\"", leftid, psk);
 	} else {
 		right = nm_setting_vpn_get_data_item (s_vpn, NM_LIBRESWAN_KEY_RIGHT);
 
@@ -1757,7 +1756,7 @@ _connect_common (NMVpnServicePlugin   *plugin,
 {
 	NMLibreswanPlugin *self = NM_LIBRESWAN_PLUGIN (plugin);
 	NMLibreswanPluginPrivate *priv = NM_LIBRESWAN_PLUGIN_GET_PRIVATE (self);
-	NMSettingVpn *s_vpn;
+	gs_unref_object NMSettingVpn *s_vpn = NULL;
 	const char *con_name = nm_connection_get_uuid (connection);
 	gs_free char *ipsec_banner = NULL;
 	gs_free char *ifupdown_script = NULL;
@@ -1795,7 +1794,8 @@ _connect_common (NMVpnServicePlugin   *plugin,
 			return FALSE;
 	}
 
-	s_vpn = nm_connection_get_setting_vpn (connection);
+	s_vpn = sanitize_setting_vpn(nm_connection_get_setting_vpn (connection),
+				     error);
 	g_assert (s_vpn);
 
 	g_object_get (self, NM_VPN_SERVICE_PLUGIN_DBUS_SERVICE_NAME, &bus_name, NULL);
-- 
2.51.0


From 55aa9f0e48acfe1ff5ee06825139bc6bdd59b4cc Mon Sep 17 00:00:00 2001
From: Gris Ge <fge@redhat.com>
Date: Fri, 17 Oct 2025 16:58:30 +0800
Subject: [PATCH 2/2] Copy secrets when sanitizing

The PSK authentication is not wkring because the
`sanitize_setting_vpn()` never copy secrets.

Fixed by introduce `PARAM_SECRET` and change `sanitize_setting_vpn()` to
copy secrets.

Signed-off-by: Gris Ge <fge@redhat.com>
---
 shared/utils.c | 46 +++++++++++++++++++++++++++++++---------------
 1 file changed, 31 insertions(+), 15 deletions(-)

diff --git a/shared/utils.c b/shared/utils.c
index 8dab313..9cb36be 100644
--- a/shared/utils.c
+++ b/shared/utils.c
@@ -38,6 +38,7 @@ enum LibreswanParamFlags {
 	PARAM_OLD	= 0x0010, /* Only include for libreswan < 4. */
 	PARAM_NEW	= 0x0020, /* Only include for libreswan >= 4. */
 	PARAM_IGNORE	= 0x0040, /* Not passed to or from Libreswan. */
+	PARAM_SECRET	= 0x0080, /* For secrets */
 };
 
 struct LibreswanParam {
@@ -347,7 +348,9 @@ static const struct LibreswanParam params[] = {
 	{ NM_LIBRESWAN_KEY_PFSGROUP,                   add,                   PARAM_IGNORE },
 	{ NM_LIBRESWAN_KEY_PSK_INPUT_MODES,            add,                   PARAM_IGNORE },
 	{ NM_LIBRESWAN_KEY_XAUTH_PASSWORD_INPUT_MODES, add,                   PARAM_IGNORE },
+	{ NM_LIBRESWAN_KEY_PSK_VALUE,                  add,                   PARAM_IGNORE | PARAM_SECRET},
 	{ NM_LIBRESWAN_KEY_PSK_VALUE "-flags",         add,                   PARAM_IGNORE },
+	{ NM_LIBRESWAN_KEY_XAUTH_PASSWORD,             add,                   PARAM_IGNORE | PARAM_SECRET},
 	{ NM_LIBRESWAN_KEY_XAUTH_PASSWORD "-flags",    add,                   PARAM_IGNORE },
 	{ NM_LIBRESWAN_KEY_NM_AUTO_DEFAULTS,           add,                   PARAM_IGNORE },
 
@@ -407,22 +410,35 @@ sanitize_setting_vpn (NMSettingVpn *s_vpn,
 		TRUE);
 
 	for (i = 0; params[i].name != NULL; i++) {
-		val = nm_setting_vpn_get_data_item (s_vpn, params[i].name);
-		if (val != NULL) {
-			handled_items++;
-		} else if (params[i].flags & PARAM_REQUIRED) {
-			g_set_error (error,
-			             NM_UTILS_ERROR,
-			             NM_UTILS_ERROR_INVALID_ARGUMENT,
-			             _("'%s' key needs to be present"),
-			             params[i].name);
-			return NULL;
-		}
-
-		if (auto_defaults) {
-			params[i].add_sanitized (sanitized, params[i].name, val);
+		if (params[i].flags & PARAM_SECRET)  {
+			val = nm_setting_vpn_get_secret(s_vpn, params[i].name);
+			if (val != NULL) {
+				nm_setting_vpn_add_secret(sanitized,
+							  params[i].name,
+							  val);
+			}
 		} else {
-			nm_setting_vpn_add_data_item (sanitized, params[i].name, val);
+			val = nm_setting_vpn_get_data_item (s_vpn,
+							    params[i].name);
+			if (val != NULL) {
+				handled_items++;
+			} else if (params[i].flags & PARAM_REQUIRED) {
+				g_set_error (error,
+					     NM_UTILS_ERROR,
+					     NM_UTILS_ERROR_INVALID_ARGUMENT,
+					     _("'%s' key needs to be present"),
+					     params[i].name);
+				return NULL;
+			}
+
+			if (auto_defaults) {
+				params[i].add_sanitized (sanitized,
+							 params[i].name, val);
+			} else {
+				nm_setting_vpn_add_data_item (sanitized,
+							      params[i].name,
+							      val);
+			}
 		}
 
 		val = nm_setting_vpn_get_data_item (sanitized, params[i].name);
-- 
2.51.0

